File: leetcode-js/134-gas-station.js
/**
 * @param {number[]} gas
 * @param {number[]} cost
 * @return {number}
 */
var canCompleteCircuit = function(gas, cost) {
    let start = 0;
    let netGas = 0;
    let curGas = 0;

    for (let i = 0; i < gas.length; i += 1) {
        netGas += gas[i] - cost[i];
        curGas += gas[i] - cost[i];
        if (curGas < 0) {
            start = i + 1;
            curGas = 0;
        }
    }

    return netGas < 0 ? -1 : start;
};

File: leetcode-java/134-gas-station.java
public class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int start = gas.length - 1;
        int end = 0;
        int sum = gas[start] - cost[start];

        while (start > end) {
            if (sum < 0) {
                start -= 1;
                sum += gas[start] - cost[start];
            } else {
                sum += gas[end] - cost[end];
                end += 1;
            }
        }

        return sum >= 0 ? start : -1;
    }
}
